### ISSUE TITLE
Singular Message Fields: invalid documentation

### ISSUE THREAD

**Issue by 3052** [protocolbuffers/protobuf#16346]
here:

https://protobuf.dev/reference/go/go-generated/#singular-message

it says:

> The compiler also generates a func (m *Concert) GetHeadliner() *Band helper function. This function returns a nil *Band if m is nil or headliner is unset. This makes it possible to chain get calls without intermediate nil checks:

this is not correct. if the method actually returned `nil`, chaining would generate a panic. instead, it likely generates a zero value pointer:

~~~go
new(Band)
&Band{}
~~~

that is not the same thing as `nil`

**Comment by 3052** [comment]
.

**Comment by 3052** [comment]
!

**Comment by 3052** [comment]
!

**Comment by 3052** [comment]
!

**Comment by honglooker** [comment]
cc @stapelberg

**Comment by stapelberg** [comment]
> https://protobuf.dev/reference/go/go-generated/#singular-message says:
> 
> > The compiler also generates a func (m *Concert) GetHeadliner() *Band helper function. This function returns a nil *Band if m is nil or headliner is unset. This makes it possible to chain get calls without intermediate nil checks:
> 
> this is not correct. if the method actually returned `nil`, chaining would generate a panic. instead, it likely generates a zero value pointer:
> 
> new(Band)
> &Band{}
> that is not the same thing as `nil`

No, the docs are correct. The generated code is (comment added for emphasis):

```go
func (m *Concert) GetHeadliner() *Band {
	if m != nil {
		return m.Headliner
	}
	return nil // returns nil *Band if *Concert m is nil
}
```

…so the code will indeed return a `nil` `*Band` (default value for a message-typed field) when `m` (a `*Concert`) is `nil`, as the docs say. 

The chaining which the docs mention:

```go
var m *Concert // defaults to nil
log.Infof("GetFoundingYear() = %d (no panic!)", m.GetHeadliner().GetFoundingYear())
```

…works because `GetFoundingYear()` is *also* performing a `nil` check (whether its `m`, a `*Band`, is `nil`) and returns the default value for an integer-typed field (`0`). That’s what the docs are trying to convey: Each part of the chain performs a nil check and returns the default.
